### ISSUE TITLE
Resolving constant fetches is unsafe

### ISSUE THREAD

**Issue by scheb** [rectorphp/rector#1816]
| Subject        | Details                                                      |
| :------------- | :----------------------------------------------------------- |
| PHP version    | PHP 7.3.7                               |
| Full Command   | vendor/composer/bin/rector process src test -vvv | 

```
parameters:
    php_version_features: '7.2'
    autoload_paths:
        - src/autoload.php
services:
    Rector\SOLID\Rector\ClassConst\PrivatizeLocalClassConstantRector: ~
    Rector\Php\Rector\ClassConst\PublicConstantVisibilityRector: ~
```

## Current Behaviour 

Some of my constants are wrongly assigned with a `private` visibility, while there was clear public usages in the code. Recognized that the parsing order actually matters. If the PHP file that declares the constant is parsed first, everything is fine. If the file(s) with the usage come first and then the declaring file at the end, the usages won't be recognized.

## Minimal PHP Code Causing Issue

The problem seems to be located in `ParsedNodesByType::collect()`, which resolves the usages on-the-fly, but at that point the file with the constant declaration hasn't necessarily been parsed yet.

Call stack (line numbers might be off, was playing around with the code):
```
#0  Rector\NodeContainer\ParsedNodesByType->matchClassTypeThatContainsConstant() called at [project/vendor/composer/rector/rector/src/NodeContainer/ParsedNodesByType.php:514]
#1  Rector\NodeContainer\ParsedNodesByType->addClassConstantFetch() called at [project/vendor/composer/rector/rector/src/NodeContainer/ParsedNodesByType.php:392]
#2  Rector\NodeContainer\ParsedNodesByType->collect() called at [project/vendor/composer/rector/rector/packages/NodeTypeResolver/src/NodeVisitor/NodeCollectorNodeVisitor.php:27]
#3  Rector\NodeTypeResolver\NodeVisitor\NodeCollectorNodeVisitor->enterNode() called at [project/vendor/composer/nikic/php-parser/lib/PhpParser/NodeTraverser.php:123]
#4  PhpParser\NodeTraverser->traverseNode() called at [project/vendor/composer/nikic/php-parser/lib/PhpParser/NodeTraverser.php:146]
```

You end up in the continue here:
https://github.com/rectorphp/rector/blob/b31b99a3aea1ba86467aa6c26ca5e971e6a1b0c0/src/NodeContainer/ParsedNodesByType.php#L446-L449
because the class declaring the constant was not loaded, not found in `$classes` so `findClassOrInterface()` returns `null` and therefore it skips the usage of the constant later here:
https://github.com/rectorphp/rector/blob/b31b99a3aea1ba86467aa6c26ca5e971e6a1b0c0/src/NodeContainer/ParsedNodesByType.php#L408
or here:
https://github.com/rectorphp/rector/blob/b31b99a3aea1ba86467aa6c26ca5e971e6a1b0c0/src/NodeContainer/ParsedNodesByType.php#L414

## Expected Behaviour

All usages of a constant should be detected, regardless of the order the files have been processed.

Would probably require some post-processing when all files have been parsed, because only then you can safely resolve the constant declaration. Not sure how to properly integrate that in rector.

In general, I think it would be wise to forbid calls to `findClass`, `findInterface` etc. from `collect()`. Doing that seems to be a bit error prone to me, because you cannot safely rely on a declaration being known while it's still parsing the code.

Let me know what you think.

**Comment by TomasVotruba** [comment]
Please provide failing php code to replicate behavior

**Comment by scheb** [comment]
Please have a look: https://github.com/scheb/rector/commit/a7533beca9d6df87cefd5322bb527587924598b4#r34594492

There's actually two conditions:
1) The constant usages must be parsed before the class declaring the constant.
2) The class declaring the constant must be a subclass.

rector is then forced find the declaring class from the class hierarchy, which it can't when the declaring class and its superclasses haven't been parsed yet.

**Comment by TomasVotruba** [comment]
I mean in the pull-request so we can see Travis fialed

**Comment by TomasVotruba** [comment]
Closed by #1822
